week-2-storage/test_parallelization.py

```diff
@@ -0,0 +1,228 @@
+    input_ids = tokens["input_ids"]
+    token_type_ids = tokens["token_type_ids"]
+    attention_mask = tokens["attention_mask"]
+    batch_size = 32
+    iterations = len(texts)//batch_size
+    with torch.no_grad():
+        for batch_idx in range(iterations):
+            if (len(texts) - iterations*batch_size) >= 0:
+                tokens_i = {
+                    "input_ids": input_ids[batch_idx*batch_size : (batch_idx+1)*batch_size],
+                    "token_type_ids": token_type_ids[batch_idx*batch_size : (batch_idx+1)*batch_size],
+                    "attention_mask": attention_mask[batch_idx*batch_size : (batch_idx+1)*batch_size],
+                }
+            else:
+                tokens_i = {
+                    "input_ids": input_ids[batch_idx*batch_size : ],
+                    "token_type_ids": token_type_ids[batch_idx*batch_size : ],
+                    "attention_mask": attention_mask[batch_idx*batch_size : ],
+                }
+            _result = model(tokens_i).cpu().numpy().tolist()
+
+
+
+if __name__ == "__main__":
+
+    nltk.download("stopwords")
+    nltk.download('wordnet')
+    nltk.download('omw-1.4')
+    stop_words = set(stopwords.words("english"))
+
+    model, tokenizer, device = init_model()
+
+    df = pd.read_csv("./dataset/train.csv")
+    # Get target column only
+    df = df["excerpt"].values.tolist()
+    print(len(df))
+    
+    # Start preprocessing and inference w\o parallelization
+    start = default_timer()
+    text_preprocessing(df, queue=None)
+    elapsed_preproc = default_timer() - start
+    print(f"[Single Process] Preprocessing: {time.strftime('%H:%M:%S', time.gmtime(elapsed_preproc))}")
+
+    start = default_timer()
+    tokens = tokeinization(tokenizer, device, df)
+    inference(model, tokens, df)
+    elapsed_inference = default_timer() - start
+    print(f"[Single Process] Inference: {time.strftime('%H:%M:%S', time.gmtime(elapsed_inference))}")
+    print(f"[Single Process] Total: {time.strftime('%H:%M:%S', time.gmtime(elapsed_inference+elapsed_preproc))}")
+
+    # #### Multiprocessing 
+
+    # Start preprocessing and inference with parallelizaiton
+    # (not the best way to architecture multiprocessing in this case)
+    cpus = multiprocessing.cpu_count()
+    print(f"Number of cpu : {cpus}")
+
+    # Preprocessing
+    jobs = []
+    queue = multiprocessing.Queue()
+    queues = []
+    dataset_size = len(df)
+    dataset_per_cpu = dataset_size // cpus
+
+    for i in range(cpus):
+        start = i*dataset_per_cpu
+        end = (i+1)*dataset_per_cpu
+        if (dataset_size - end) >= dataset_per_cpu:
+            p = multiprocessing.Process(target=text_preprocessing, args=(df[start : end], queue))
+            jobs.append(p)
+            queues.append(queue)
+        else:
+            p = multiprocessing.Process(target=text_preprocessing, args=(df[start : ], queue))
+            jobs.append(p)
+            queues.append(queue)
```

Comment:
curious why you did not get any speedup, what if we increase dataset size? 


what is reasoning to Queue btw? should be process pool easy to work with

Reply:
Data processing has 3 steps:
- preporcessing (parallelization dropped down execution time from 5 to 1 second)
- tokenization (managed by pytorch and I didn't manage to parallelize it even with torch multiprocessing)
- inference (actually call the model with tokens to predict scores, parallelization is done by batching mechanism)

So, in this particular setup I can parallelize only preprocessing stage.

A few words about Queue. I know that it is better to use more high level abstractions like ProcessExecutor, but I just used to use Processes with Queues.